### PR TITLE
Explicit never early terminates

### PR DESCRIPTION
**Pull request by b1rdex**
Closes https://github.com/phpstan/phpstan/issues/2638

### CODE REVIEW

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
This is almost perfect PR :) Please test it in a more modern way - see `NodeScopeResolverTest::testFileAsserts` - add a new dataProvider in it. You can use `assertVariableCertainty` function inside the test file. Thanks.

**Comment by b1rdex** [comment]
@ondrejmirtes I updated the test

File: tests/PHPStan/Analyser/NodeScopeResolverTest.php [hunk 1]
```diff
@@ -10259,6 +10259,12 @@ public function dataBug4016(): array
 		return $this->gatherAssertTypes(__DIR__ . '/data/bug-4016.php');
 	}
 
+	/** @return array<string, mixed[]> */
+	public function phpDocNeverEarlyTerminates(): array
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/NodeScopeResolverTest.php:10263]: Needs to start with `data` as any other data provider here.

File: tests/PHPStan/Analyser/data/early-termination-phpdoc.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+namespace EarlyTermination;
+
+use PHPStan\TrinaryLogic;
+use function PHPStan\Analyser\assertVariableCertainty;
+
+function(): void {
+	$pre = 1;
+	Foo::doBarPhpDoc();
+	$a = 1;
+	assertVariableCertainty(TrinaryLogic::createYes(), $pre);
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/early-termination-phpdoc.php:12]: I'm not even sure if these are executed. They might not be. Have you seen them fail?

The typical test here should be this instead:

```php
if (rand(0, 1)) {
    $a = 1;
} else {
    Foo::doBarPhpDoc();
}

assertVariableCertainty(TrinaryLogic::createYes(), $a);
```

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by ondrejmirtes** [comment]
Yeah, I just verified that the asserts are not verified because they're "dead code". I'll bring it over the finish line myselfů

**Comment by ondrejmirtes** [comment]
Thank you. There needs to be a second part - verifying that a function with `@return never` actually never returns.

**Pull request merged**